docs/studies/hidden_linear_function.ipynb

```diff
@@ -0,0 +1,540 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Hidden Linear Function Problem\n",
+    "\n",
+    "In this notebook we consider a problem from the paper \"[Quantum advantage with shallow circuits](https://arxiv.org/abs/1704.00690)\" and build a quantum circuit, which solves it, in Cirq. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Introduction\n",
+    "\n",
+    "### The problem\n",
+    "\n",
+    "In this problem we consider a [quadratic form](https://en.wikipedia.org/wiki/Quadratic_form) of a [binary vector](https://en.wikipedia.org/wiki/Bit_array) and with binary coefficients (but additions and multiplications are evaluated modulo 4). Then we restrict this quadratic form, i.e. we allow to use only certain binary vectors as input. It turns out that under this restriction this quadratic form is equivalent to a linear function, i.e. it just evaluates dot product of input vector and certain scalar vector. Task is to find this scalar vector.\n",
+    "\n",
+    "In other words, we have linear function, which is \"hidden\" inside a quadratic form.\n",
+    "\n",
+    "### Formal statement of the problem\n",
+    "\n",
+    "Consider $A \\in \\mathbb{F}_2^{n \\times n}$ - upper-triangular binary matrix of size $n \\times n$, $b \\in \\mathbb{F}_2^n$ - binary vector of length $n$.\n",
+    "\n",
+    "Define a function $q : \\mathbb{F}_2^n \\to \\mathbb{Z}_4 $:\n",
+    "\n",
+    "$$q(x) = (2 x^T A x + b^T x) ~\\text{mod}~ 4, $$ \n",
+    "\n",
+    "Also define\n",
+    "\n",
+    "$$\\mathcal{L}_q = \\Big\\{x \\in  \\mathbb{F}_2^n : q(x \\oplus y) = (q(x) + q(y)) ~\\text{mod}~ 4 ~~ \\forall y \\in \\mathbb{F}_2^n \\Big\\}.$$\n",
```

Comment:
I think we should add more explanation into this section.  What is the motivation behind these definitions?  Can we give the reader some intuition behind the math?

Reply:
This problem is indeed contrived. I don't know how exactly authors of the paper came up with this particular problem statement. So I don't know what to write here. This problem doesn't represent any practcal problem from real life.

Reply:
Can we give any intuition to the reader beyond a bunch of math that they need to read through?  I think it would be easier if we at least name the variables, e.g.  z is the hidden solution, etc.  That way, the reader can get a feel for what A, b, q, L_q, and z are.

Reply:
I added verbal description for all variables in the end of this section.